### ISSUE TITLE
Photo view: load new images when reaching bottom

### ISSUE THREAD

**Issue by tobiasKaminsky** [nextcloud/android#1901]
Currently photo view tries to load all images that are stored on server and this results in very long waiting.
@rullzer had the idea to show first only the latest 7 days and when reaching the bottom of the list, then show/add again 7 days.
This would give the user at the beginning a very fast experience...

@mario it should be possible to limit the search to a certain timespan?

**Comment by andyxh** [comment]
Well at first I think all the phones should take a long time to load but once they’ve all loaded there should be a solution that the app caches some of the data and remember the phones that where added, then if you add a new phone it will show up on the photos tab too do you understand what I’m saying?

**Comment by tobiasKaminsky** [comment]
The main problem is that displaying > 5000 images is just taking too long on internal android computing of the layout, resizing thumbnails, ...
Unfortunately this cannot be stored.

Retrieving all the infos from the server also takes some time, but is not that much in comparison.

**Comment by andyxh** [comment]
Correct but the iOS system is doing it

**Comment by rullzer** [comment]
@andyxh this won't work in several setups.

* If you have a huge photo collection (I have a easily a few hunderd GB of pictures) the initial request will never finish. It will just time out.
* How will my phone find out there are new pictures that are uploaded for example with the desktop client? We'd have to query the server again. But we can't tell it just give us new stuff. Since the desktop client preserves mtime. So if you upload pictures from a month ago they will be sorted a month ago.

Long story short. Date range requests are an easy way to limit the request scope. While still providing the feeling of infinite scroll to the users.

**Comment by andyxh** [comment]
Well when you click on the photos folder and when your scrolling throw the phone the picture show up and it keeps it cached, now can you take that cached data and show it in the photos tab ?

**Comment by zippy-man** [comment]
How about a library to help load images in a better manner. I have used this before in a previous project. Not sure if it could help here as I have not reviewed any nexcloud code for image loading. 

https://github.com/bumptech/glide

**Comment by andyxh** [comment]
iOS has a photos tab option and once you click on the photos album and it loads all the photo tub nails you can go back into photos tab and it will show you all the photos. Not sure why android cant make something like this happen or talk to the iOS team on how they developed that feature.

Sent from my iPhone 7 Plus

On Jan 5, 2018, at 11:23 PM, zippy-man <notifications@github.com<mailto:notifications@github.com>> wrote:


How about a library to help load images in a better manner. I have used this before in a previous project. Not sure if it could help here as I have not reviewed any nexcloud code for image loading.

https://github.com/bumptech/glide<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbumptech%2Fglide&data=02%7C01%7C%7C00746e822b0b4905f0c808d554c5961d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636508129957260471&sdata=%2BX2FyfP62JnvV5vz5GmUgAiz%2BUN7ASxOHiQhFYEyV8Y%3D&reserved=0>

**Comment by mario** [comment]
Actually, I'm the one who helped the iOS team with Photos tab (and search in general). iOS and the Android are different platforms, starting off a different maturity codebase.

The idea outlined by @rullzer is a good one, and we'll use that. This particular thing + recyclerview will speed up quite a lot of things.

**Comment by andyxh** [comment]
Well thank you for all your hard work! 🙂 yes, I also agree that it’s a good idea.

Sent from my iPhone 7 Plus

On Jan 6, 2018, at 1:46 AM, Mario Đanić <notifications@github.com<mailto:notifications@github.com>> wrote:


Actually, I'm the one who helped the iOS team with Photos tab (and search in general). iOS and the Android are different platforms, starting off a different maturity codebas@rullzer<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frullzer&data=02%7C01%7C%7C7dbab1aa36fa422c18b608d554d9995c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636508215910727061&sdata=Bh6prnp0upFO32vTQHdQhtZg6mJvu5PUMl7ltYnFxz4%3D&reserved=0>e.

The idea outlined by @rullzer<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frullzer&data=02%7C01%7C%7C7dbab1aa36fa422c18b608d554d9995c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636508215910727061&sdata=Bh6prnp0upFO32vTQHdQhtZg6mJvu5PUMl7ltYnFxz4%3D&reserved=0> is a good one, and we'll use that. This particular thing + recyclerview will speed up quite a lot of things.

**Comment by tobiasKaminsky** [comment]
@mario @AndyScherzinger 
I implemented this now for photo search: 
- it first loads 3 x screen, e.g. 3 x 5 rows x 4 columns = 60 items
- when reaching the bottom, it loads the next 60 items

This is possible as photo view is sorted by time and on next items I can limit the search and find only items that are older than the last timestamp.
My idea was to use the dynamic fetching also for our other search implementations, but this might be trickier for "pagination":
- sort by size would work
- sort by time: see above
- sort by name: there is afaik no way for comparing the file name --> this would be needed for a "regular" file search.

**Comment by rullzer** [comment]
Sort by name is kind of weird also I would have no idea how the query to search for the rest files would look like 

Sort by time probably makes sense... Most recent first (or least recent).

**Comment by AndyScherzinger** [comment]
Fine by me @tobiasKaminsky simply because the only truly working solution is block-reading because time doesn't work. If I went on vacation "a day" can mean hundred of pictures taken with 2 phones and one DSLM. So always retrieving n elements seems to be the way to go.

**Comment by tobiasKaminsky** [comment]
@rullzer then we need for a file search (e.g. you type "test") and showing only 1-50 results, then 51-100 results a support for real pagination...

**Comment by rullzer** [comment]
Well that is something much harder to do. Or it isn't hard but hard to do efficiently.

**Comment by mario** [comment]
Cc @icewind :)

**Comment by rbu** [comment]
@tobiasKaminsky's PR is #2250.

**Comment by rcdailey** [comment]
Any update on this since May? This is absolutely killing me. #1586 was closed in favor of this PR, but I'm worried that because the issue was closed without an actual fix merged in, we're losing visibility into this issue.

This issue has been around for over a year. Is this just not a priority for the team? I guess not many people care about photos or have big enough photo collections for this to matter?

**Comment by tobiasKaminsky** [comment]
I asked server guys on #2250.

**Comment by via-justa** [comment]
What about generating small previews that the photos gallery will fetch and creating a GET request for the full image when clicking on a pic? Wouldn't it speed thing dramatically? I don't think that the gallery needs pics bigger than 20-30ppi. The preview generation can be done on the server as a cron job and/or when uploading a new photo. I think that the fast delivery worth the disk space cost.
If combining it with the idea in this issue I'm sure it will be super fast loading time

**Comment by tobiasKaminsky** [comment]
The problem is the search on server for now. 
If you have 50k files, and 2k images then the search on server takes ~30s.
Then all 2k infos get transferred to client (~2-3mb) and parsed and only after that the previews are downloaded.

**Comment by e-alfred** [comment]
We are now at Nextcloud 16, is this already resolved on the server side?
